content/blog/2022-04-19-tracing/index.md

```diff
@@ -0,0 +1,142 @@
+
+The author also introduces an intresting optimization technique called Tree Stiching. The purpose of tree stitching is to reuse activation record when a side exit is met. The reason is fairly straightforward, since it's easy to see that within a loop, the register allocation and memory layouts are often invariant. the exiting trace only needs to write live
+register-carried values back to its trace activation record. However, with multi-code processors being the majority (unlike when this paper just got published), it's probably just easier to recompile the entire trace-tree in backend.
+
+Additional optimization techniques when a LIR trace gets compiled to machine code as a set of filters, including
+1. Float to Integer Conversion
+2. Constant Subexpression Elimination
+3. Constant Folding
+4. Dead Code Elimination
+5. Dead Stack Elimination
+
+# Evaluation
+The evaluation of is paper is performed on a Macbook Pro over SunSpider, which is a standard testing suite for JavaScript with a variety of applications. The experimental environment is very reasonable to me, since it's a very good representation of JavaScript workloads and user's device. The author benchmarks TraceMonkey against SpiderMonkey (baseline interpreter), SFX (inline threading interpreter in Apple's Webkit), V8 (Method-based JIT developed by Google).
+
+Based on the figure above, traceMonkey is the fastest VM on 9 out of 26 benchmarks. In particular, TraceMonkey performs on an average of ~5x compared to SFX and V8 on `bitops` benchmarks, which are mainly short bitwise operations. In addition to that, TraceMonkey performs well in maths and crypto, slightly edging SFX and V8. TraceMonkey does not achieve the same level of speedup against SFX and V8 on programs with a high number of traces or tiny loop bodies, as well as cases where TraceMonkey currently not tracing (i.e. recursion), although I believe it's relatively easy and reasonable to implement tracing for recursion.
+
+The detailed trace recording statistics provides more insights into the performance. One thing I noticed is that Trace/Loop is inversely correlated with Speedup. The reason I believe is that more traces per loop means higher compilation costs, and also higher costs to call inner loop / branch traces. This shows that even with many optimization techniques implemented upon trace trees, dealing with nested loops and loops with multiple branches are still the main challenges.
+
+# Discussion Question
+- Why tracing-JITs cannot compete with method-JITs? Even though TraceMonkey was successful at its time, it was eventually replaced by a method-JIT.
+- Would another definition of 'hot' helps with tracing-JITs performance?
+- What are the biggest contributors to modern JavaScripts' speedup?
```

Comment:
I would actually recommend not including the questions themselves here. Instead, you can include brief summaries of your answers (perhaps synthesizing the class's answers), if that seems feasible.